gflownet/utils/batch.py

```diff
@@ -94,131 +166,164 @@ def add_to_batch(
         valids : list
             A list of boolean values indicated whether the actions were valid.
 
-        masks_invalid_actions_forward : list
-            A list of masks indicating, among all the actions that could have been
-            selected for all environments, which ones were invalid. Optional, will be
-            computed if not provided.
+        backward : bool
+            A boolean value indicating whether the action was sampled backward (False
+            by dfefault). If True, the behavior is slightly different so as to match
+            what is stored in forward sampling:
+                - If it is the first state in the trajectory (action from a done
+                  state/env), then done is stored as True, instead of taking env.done
+                  which will be False after having performed the step.
+                - If it is not the first state in the trajectory, the stored state will
+                  be the previous one in the trajectory, to match the state-action
+                  stored in forward sampling and the convention that the source state
+                  is not stored, but the terminating state is repeated with action eos.
 
         train : bool
             A boolean value indicating whether the data to add to the batch will be used
             for training. Optional, default is True.
         """
-        if self.is_processed:
-            raise Exception("Cannot append to the processed batch")
-
-        # Sample masks of invalid actions if required and none are provided
-        if masks_invalid_actions_forward is None:
-            if train:
-                masks_invalid_actions_forward = [
-                    env.get_mask_invalid_actions_forward() for env in envs
-                ]
-            else:
-                masks_invalid_actions_forward = [None] * len(envs)
+        # TODO: do we need this?
+        if self.continuous is None:
+            self.continuous = envs[0].continuous
 
         # Add data samples to the batch
-        for sample_data in zip(envs, actions, valids, masks_invalid_actions_forward):
-            env, action, valid, mask_forward = sample_data
-            self.envs.update({env.id: env})
+        for env, action, valid in zip(envs, actions, valids):
+            if train is False and env.done is False:
+                continue
             if not valid:
                 continue
-            if train:
-                self.states.append(deepcopy(env.state))
-                self.actions.append(action)
-                self.env_ids.append(env.id)
+            # Add env to dictionary
+            if env.id not in self.envs:
+                self.envs.update({env.id: env})
```

Comment:
If the number of envs is potentially large (say in the 100+ range) I'd suggest using tuples here, which are much more memory efficient.

https://www.geeksforgeeks.org/python-memory-consumption-dictionary-vs-list-of-tuples/

Reply:
This comment goes for any place you index a dict with `env.id`

Reply:
uhm ok good to know, thanks! Adding an issue.